boa/src/syntax/parser/cursor/buffered_lexer/mod.rs

```diff
@@ -0,0 +1,230 @@
+    ///
+    /// It will not fill two line terminators one after the other.
+    fn fill(&mut self) -> Result<(), ParseError> {
+        debug_assert!(
+            self.write_index < PEEK_BUF_SIZE,
+            "write index went out of bounds"
+        );
+
+        let previous_index = self.write_index.checked_sub(1).unwrap_or(PEEK_BUF_SIZE - 1);
+
+        if let Some(ref token) = self.peeked[previous_index] {
+            if token.kind() == &TokenKind::LineTerminator {
+                // We don't want to have multiple contiguous line terminators in the buffer, since
+                // they have no meaning.
+                let next = loop {
+                    let next = self.lexer.next()?;
+                    if let Some(ref token) = next {
+                        if token.kind() != &TokenKind::LineTerminator {
+                            break next;
+                        }
+                    } else {
+                        break None;
+                    }
+                };
+
+                self.peeked[self.write_index] = next;
+            } else {
+                self.peeked[self.write_index] = self.lexer.next()?;
+            }
+        } else {
+            self.peeked[self.write_index] = self.lexer.next()?;
+        }
+        self.write_index = (self.write_index + 1) % PEEK_BUF_SIZE;
+
+        debug_assert_ne!(
+            self.read_index, self.write_index,
+            "we reached the read index with the write index"
+        );
+        debug_assert!(
+            self.read_index < PEEK_BUF_SIZE,
+            "read index went out of bounds"
+        );
+
+        Ok(())
+    }
+
+    /// Moves the cursor to the next token and returns the token.
+    ///
+    /// If skip_line_terminators is true then line terminators will be discarded.
+    ///
+    /// This follows iterator semantics in that a `peek(0, false)` followed by a `next(false)` will
+    /// return the same value. Note that because a `peek(n, false)` may return a line terminator a
+    // subsequent `next(true)` may not return the same value.
+    pub(super) fn next(
+        &mut self,
+        skip_line_terminators: bool,
+    ) -> Result<Option<Token>, ParseError> {
+        if self.read_index == self.write_index {
+            self.fill()?;
+        }
+
+        if let Some(ref token) = self.peeked[self.read_index] {
+            let tok = if !skip_line_terminators || token.kind() != &TokenKind::LineTerminator {
+                mem::replace(&mut self.peeked[self.read_index], None)
+            } else {
+                // We only store 1 contiguous line terminator, so if the one at `self.read_index`
+                // was a line terminator, we know that the next won't be one.
+                self.read_index = (self.read_index + 1) % PEEK_BUF_SIZE;
+                if self.read_index == self.write_index {
+                    self.fill()?;
+                }
+
+                mem::replace(&mut self.peeked[self.read_index], None)
```

Comment:
I think we should replace these by `take()`. We don't have benchmarks to decide otherwise, and it would be more clear for other developers.